### PR TITLE
Changes related to restG4 MT

### PR DESCRIPTION
**Pull request by lobis**
lobis Large: 2076 [](https://gitlab.cern.ch/rest-for-physics/geant4lib/-/commits/lobis-prepare-for-multithreading) [](https://gitlab.cern.ch/rest-for-physics/restG4/-/commits/lobis-prepare-for-multithreading) [](https://gitlab.cern.ch/rest-for-physics/framework/-/commits/lobis-prepare-for-multithreading)

* https://github.com/rest-for-physics/restG4/pull/58

## Data Containers

Data containers (Event, Track, Steps) have been updated: some member names have been changed for consistency and others deemed redundant have been removed (such as `TRestGeant4Tracks::fNTracks`, same as `fTracks.size()` etc.). Many unused or redundant methods have also been removed for simplicity. Class version have been updated.

Data containers now have references to other containers as members. This references are not stored on the root file (`//!`) and are initialized dynamically by `TRestRun::GetEntry`. The allow things like accessing the event from a track, getting the pointer to the parent track directly from the child track etc.

Data containers initialization is handled by `restG4` with a constructor taking a const reference to the corresponding geant4 construct as argument. `Geant4lib` purpose is to perform analysis on simulation events, we don't need to create the data using `Geant4lib`.

The way subevent primary information is stored has been rethinked:

Now you can access the subevent primary information (position, energy, particle...) and the event primary information. For example for a nuclear decay subevent you can get the position of the primary decay particle via `GetSubEventPrimaryEventOrigin` and access the original primary position via `GetEventPrimaryEventOrigin`.

Also the member `fSubEventPrimaryParticleName` automatically serves to store the decay particle name (if it exists) so we no longer need to use the event tag to store this information.

## Metadata

`TRestGeant4Metadata` has been simplified and now longer directly contains primary generator information. This is now contained in the new class `TRestGeant4PrimaryGeneratorInfo`. The RML declaration remains the same. The only change is regarding the getters for generator info: previously you could do things like `TRestGeant4Metadata::GetGeneratorPosition()`, now you need to do something like `TRestGeant4Metadata::GetGeant4PrimaryGeneratorInfo().GetGeneratorPosition()`.

`TRestGeant4PrimaryGeneratorInfo` contains the same logic previously present on `TRestGeant4Metadata` and some other functions have been implemented. Better validation using enums to make sure a correct option has been selected, and to store a consistent string name for the options.

`energyDist` and `angularDist` rml keywords have been replaced by `energy` and `angular`. The previous names also work so both are supported and this should not break anything. All examples have been updated.

A new type of distribution has been introduced called `Formula` which can be used on angular and energy distributions. Currently it only supports predefined formulas but it can support used defined formulas with little change.

For example: `<angular type="TH1D" file="CosmicAngles.root" spctName="Theta2" direction="(0,0,-1)"/>` now can also be `             <angular type="formula" name="Cos2" direction="(0,0,-1)"/>` with does not require referencing a file. It should also have more precission since its sampling the mathematical formula directly. The `TF1` is stored in the particle source and can be retrieved and plotted directly.

In the case of energy distributions they are defined in a predefined range (one where the formula is valid) and the user can choose a different range with the usual `range=` xml syntax. A validation logic has been introduced to make sure that the final range is always the intersection with the formula validity range, so we cannot use a energy distribution ranging from 1 MeV to 100 MeV to sample from 100 keV to 20 MeV for example. The final range would be from 1MeV to 20MeV automatically without giving error (the range can be consulted on the metadata as usual).

### CODE REVIEW

**Comment by lobis** [comment]
Currently the pipeline is failing due to the reference file of PandaX pipeline, I could not regenerate it. More info on https://github.com/rest-for-physics/framework/pull/273.

File: inc/TRestGeant4Metadata.h [hunk 1]
```diff
@@ -186,27 +114,46 @@ class TRestGeant4Metadata : public TRestMetadata {
...
     /// deposit a non-zero energy on this volume will be registered.
-    TString fSensitiveVolume;
+    std::vector<TString> fSensitiveVolumes;
 
     /// The number of events simulated, or to be simulated.
-    Int_t fNEvents;
+    Int_t fNEvents = 0;
+
+    /// The number of events the user desired to be on the file
+    Int_t fNDesiredEntries = 0;
```
Comment (juanangp) [inline inc/TRestGeant4Metadata.h:131]: `fNRequestedEntries`

File: inc/TRestGeant4PrimaryGeneratorInfo.h [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+enum class SpatialGeneratorShapes {
+    GDML,
+    WALL,
+    CIRCLE,
+    BOX,
+    SPHERE,
+    CYLINDER,
+};
+
```
Comment (juanangp) [inline inc/TRestGeant4PrimaryGeneratorInfo.h:31]: `constexp std:map<SpatialGeneratorShapes, std::string> SpatialGeneratorTypesMap`

File: inc/TRestGeant4Metadata.h (hunk 1 above)
Comment (lobis) [reply inc/TRestGeant4Metadata.h:131]: updated

File: inc/TRestGeant4PrimaryGeneratorInfo.h (hunk 2 above)
Comment (lobis) [reply inc/TRestGeant4PrimaryGeneratorInfo.h:31]: I agree we should revisit this but I will do it in a separate PR, feel free to implement the changes if you want.

**Review by juanangp — APPROVED** [review]

**Comment by juanangp** [comment]
Remember to change the pipeline `@master`

**Pull request merged**